Update MBC3.md

nummacway opened the pull request:
Results from CasualPokePlayer and me, as discussed in the Research channel on Discord on Tuesday. Also make RAM bank count and RAM size consistent and add some more details how it handles wirting invalid values.

nummacway committed:
Update MBC3.md

Results from CasualPokePlayer and me, as discussed in the Research channel on Discord on Tuesday. Also make RAM bank count and RAM size consistent and add some more details how it handles wirting invalid values.

ISSOtm commented on src/MBC3.md:
```diff
@@ -24,6 +24,8 @@ Depending on the current Bank Number/RTC Register selection (see below),
 this memory space is used to access an 8 KiB external RAM Bank, or a
 single RTC Register.
 
+The Japanese version of Pokémon Crystal Version is the only official game to have an MBC3 with 8 RAM banks (for a total of 64 KiB). It is sometimes referred to as MBC30, reflecting the print on the chip, although the cartridge type in the header is not different.
```
This should be moved under the initial description paragraph.

ISSOtm commented on src/MBC3.md:
```diff
@@ -58,11 +60,23 @@ Controls what is mapped into memory at A000-BFFF.
+
+The exact behavior of this register varies depending on hardware:
+
+MBC3B provides a running clock on power-on and after writing any even value to this register. It is still recommended to latch the clock by writing any odd value. MBC3B can only latch while it provides a running clock, so you must write an even value before you can write an odd value again.
```
I don't think the two versions of this chip are mentioned before this point? I'd suggest adding a paragraph explaining this in the intro section.

ISSOtm reviewed (changes requested):
Thank you for the contribution!

nummacway commented on src/MBC3.md:
```diff
@@ -73,8 +87,11 @@ clock keeps ticking.
+
+Bits that are not required to store the above information will be ignored and always read 0.
+
+You can write values larger than the ones mentioned above (up to 63 for seconds and minutes, and up to 31 for hours). Invalid values will then continue incrementing like a valid value and will only overflow once the available bits no longer suffice. This overflow however will not cause a carry, neither does writing 60 or 24 directly. For example, if you write 30:59:63 (and clear the Halt Flag), i ...
```
So this behavior is thought to be universal and the last sentence should be removed?

nummacway commented:
We also might want to add the following, which is true on MBC3B, but @CasualPokePlayer please confirm on MBC3A:
* You cannot see your writes to the counter registers reflected while the clock is latched.
* When writing to the counter registers, both latch and halt are optional.

nummacway committed:
Accept suggestions from code review

Co-authored-by: Eldred Habert <me@eldred.fr>

nummacway committed:
Update MBC3.md

Reflecting suggestions from the discussion

nummacway commented:
Before we merge, we should agree on a consistent capitalization scheme. It is entirely random right now. Looks like it has originally been written by a German.

quinnyo reviewed (changes requested):
Thank you for your efforts!

I don't think it's quite right to change the supported ROM/RAM size in this way. It seems misleading to say plainly that MBC3 supports 4 MiB ROM / 64 KiB RAM, when only one special variant actually does. MBC30, insofar as it's part of the "MBC3 family", is an anomaly, not a representative.

To aid review and future changes, it's helpful to spread the text over multiple lines. The easiest/obvious place to do this is per-sentence, but it can make sense to breaking up long sentences in some cases. Note that the old text has hard-coded line wrapping, which is undesirable.

nummacway committed:
Update MBC3.md

Generally avoided the term MBC3 when not speaking about a certain variant

quinnyo commented on src/MBC3.md:
```diff
@@ -58,11 +62,25 @@ Controls what is mapped into memory at A000-BFFF.
+
+MBC3A's clock counters are indeterminate by default. Writing any value to this register latches the clock. MBC3A cannot provide a running clock. Naturally, it can latch repeatedly.
+
+**tl;dr:** Write $00 then $01 to this register to safely trigger latching on all versions of the chip.
```
It seems that all of my comments in the review ended up pointing at more lines than what I anchored them to.
The line I was referring to here was only the "tl;dr" one at the bottom, in case that wasn't clear.

nummacway committed:
Update MBC3.md

- Made clearer that the "default state" of the MBC3A refers to the state when powered on
- Undid change requested by @issotm based on feedback from others and my own opinion
- Changed "RTC registers" to "clock counter registers" (since 4000-5FFF is an RTC register, but is not what the term was referring to)
- Unified capitalization: lowercase in text (those are no proper nouns!) and title case in titles
- Unified line breaks (infinte length)

avivace commented on src/MBC3.md:
```diff
@@ -58,11 +62,25 @@ Controls what is mapped into memory at A000-BFFF.
+
+MBC3B provides a running clock on power-on and after writing any even value to this register. It is still recommended to latch the clock by writing any odd value. MBC3B can only latch while it provides a running clock, so you must write an even value before you can write an odd value again.
+
+MBC3A's clock counters are indeterminate by default. Writing any value to this register latches the clock. MBC3A cannot provide a running clock. Naturally, it can latch repeatedly.
```
I believe we use expressions such as "These registers are left entirely uninitialized" instead of 'indeterminate'

nummacway commented:
Sorry, I misclicked you @quinnyo.

Anyway, where are we stuck here? PRs in GBDEV projects usually result in an endless discussion about odds and ends, when the original change would already be a significant improvement.

ISSOtm commented:
Largely, lack of momentum and maintainer time, I guess. Sorry, I'm stretched fairly thin, personally.

ISSOtm commented:
Ah, also, I see that several of my review comments have been neither responded to nor marked as resolved, so I assumed that the author still had them pending. What's their status?

quinnyo commented:
I'd like to see a lot of the changes here merged, but I don't think this PR should be merged in its current state. There's unresolved review comments, and the MBC30 issue I mentioned above.

To elaborate on what I said above:

I consider the MBC30 to be a different mapper to MBC3.
It does not make sense to state the capabilities of MBC30 as that of MBC3 or vice versa.
Different capabilities = different mapper.

- The 'A' and 'B' variants look obviously like revisions. i.e. an update to address some issue. Each revision replaced the one that came before. Games shipped with MBC3 first, then MBC3A, then MBC3B. The capabilities remained the same.
  - MBC30 did not succeed any of the others in the same way. The bulk of MBC3B titles shipped after MBC30 started.
- MBC30 supports more memory. Software taking advantage of the MBC30's capabilities cannot possibly run on the MBC3.
- There was at least one PCB revision introduced with MBC3A and MBC3B was installed on that same board. MBC30 had a special board just for it.
- The *only* reason to say MBC30 is MBC3 is the ID in the cartridge header, which we know is technically meaningless.

It would be good to document the MBC30, and I think it's fine to do that on the MBC3 page, given the close relationship between them. But MBC30 should be presented as a special one-off variant, not a representative or archetype of MBC3.

nummacway commented:
> No matter how many improvements you made and how important they are, we can't merge your PR until _you_ make sure that everything is 100% perfect.

This attitude and the resulting tedious process is too exhausting to me. I'm extremely sorry do not have the energy for this. I'm used to wikis where people ask themselves only a single yes-or-no question: "Did the change improve the article?" And if people think that something should be changed in addition to that, _people_ do this.

ISSOtm commented:
My apologies that it's come to this.

We're walking a fine tradeoff here: Pan Docs *is* an authoritative source, and thus we hold a high quality standard. This benefits all readers since they know they can trust what's written, at high cost to contributors. A free-edit wiki would make editing simpler, but then put untenable burden on any moderators seeking to verify whether claims are correct.

Put another way, we aim to have a similar level of backing as Wikipedia requires sources. But, instead of sources, we rely on test ROMs and other such works.

We're trying to bridge technical knowledge and pedagogy. You know what's to be documented, we know how to convey it (or, at least, both sides are fairly confident in that), and we're seeking a way for The Info to cross that gap. Probably our contributing model is broken; but I'm not sure what an appropriate replacement would be.

avivace commented:
As @ISSOtm noted, Pan Docs holds itself to certain standards around what gets merged and how. Part of that process is surfacing planned additions to the community early and gathering actionable feedback. In that regard, the discussions in this thread have been well-targeted and raised some interesting questions.

This PR touched on a number of sensitive areas that sparked some genuinely valuable discussion and engagement from multiple contributors. If you don't have the bandwidth to follow up on the review process, that's completely fine - @nummacway please don't feel obligated just because you were the original author. Maintainers and other contributors are happy to carry it forward.

I'm going to re-open this PR to not lose the various threads, please feel free to unsubscribe from notifications / leave the conversation @nummacway .

pinobatch commented:
I offer something to consider with regard to a replacement model: Some Wikimedia projects, such as Wikibooks, have a ["pending changes" mechanism](https://en.wikibooks.org/wiki/Help:Tracking_changes#Reviewing_pages). Page revisions by editors who aren't established, regular editors are marked "pending", and anonymous visitors see the latest version of a page that isn't marked pending.

nummacway commented:
I'm trying to help. Research already took some hours of testing (not just for me, but also CasualPokePlayer), plus another hour or two to edit the file, because I gave my best to meet the required standards. It makes me sad that this PR then ends up in a seesaw just because of odds-and-ends – requested changes that (at least in a wiki) the requesting user could more easily have done themselves than making a post about it. There _are_ proposed changes in GitHub, but in the end, they **place the load on the original author of the PR**. This whole concept makes sense for software (which git is designed for) - but it doesn't make sense for text, especially if writing about _facts_. The additional load placed on the author of the PR in this case is unreasonable and does not improve the result (although such result might take a few – but fast – iterations (edits) on a wiki).

After posting my message above, I also came up with the "reliable sources" shortcoming of my argumentation, and that Wikipedia editors will – in fact – revert your change if it lacks sources. This is partly because original research is banned on Wikipedia. However, Pan Docs consists (almost) exclusively of original research (the "almost" is for the case that Liji's original research is quoted instead). How we got to our results is verifyable via Discord. And the requested changes were unrelated to the actual research.

If your edit on Wikipedia is correct, provides sources and only lacks formatting or related standards (or if your English just isn't perfect), other Wikipedia editors will more likely fix this _themselves_ instead of just reverting the change (which is their way of rejecting a PR, because wikis auto-merge all changes into the source – they might just not display unreviewed versions to anonymous users). Alternatively, or if your edit did help but is incomplete, they will add a stub template. (This is what I was trying to achieve with the "help wanted" box.) From my own experience as the owner of a medium-sized wiki (20k articles), a wiki is _very_ likely to end up with a version of the article that everyone is fine with. Again, we're writing about _facts_ that noone can deny once a test ROM and a photo of its output is available. The rest is just style.

Pan Docs has always claimed high standards, but is it? When making my cheat sheets for my own use, I found some things that were inconsistent between several sources or that I didn't understand. In both situations, I did my own testing, wrote my own test ROMs and so on. I did find errors, but I find the editing process of Pan Docs too bureaucratic to post them.

avivace commented on src/MBC3.md:
```diff
@@ -15,80 +16,81 @@ Contains the first 16 KiB of the ROM.
+
+The exact latching behavior of MBC3 and MBC30 has not been tested and the sample size could be improved even for the MBC3A and MBC3B.
+
+If you would like to help, have a flashcart and any official RTC cartridge, please reach out to us on gbdev Discord so you can be given the test ROMs.
```
Can such test ROM be made public? @CasualPokePlayer @nummacway ?

avivace replied:
Author of test rom @CasualPokePlayer replied: https://github.com/CasualPokePlayer/test-roms/tree/latch-rtc-test

avivace commented:
> After posting my message above, I also came up with the "reliable sources" shortcoming of my argumentation, and that Wikipedia editors will – in fact – revert your change if it lacks sources. This is partly because original research is banned on Wikipedia. However, Pan Docs consists (almost) exclusively of original research (the "almost" is for the case that Liji's original research is quoted instead). How we got to our results is verifyable via Discord. And the requested changes were unrelated to the actual research.

Discord is to be avoided as a place to hold important and/or supporting information. See the comment also on the 'research needed' redirecting to discord.

> Pan Docs has always claimed high standards, but is it? When making my cheat sheets for my own use, I found some things that were inconsistent between several sources or that I didn't understand. In both situations, I did my own testing, wrote my own test ROMs and so on. I did find errors, but I find the editing process of Pan Docs too bureaucratic to post them.

Please do report the errors opening a ticket (an Issue) and (optionally) mentioning your suggestions/corrections

avivace commented:
Is this superseded by #661 ?

nummacway commented:
No. If any, this PR superseeds #661.

ISSOtm commented:
Fair, but since that PR was merged, we'll need to rebase this one to reconcile the two.
